File: janitor-core/src/main/java/com/netflix/spinnaker/janitor/model/Rule.java
/*
 * Copyright 2017 Netflix, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.netflix.spinnaker.janitor.model;

/**
 * The rule implementing a logic to decide if a resource should be considered as a candidate of cleanup.
 */

public interface Rule extends Comparable<Rule> {

  /**
   * Getter for rule name
   * @return rule name
   */

  String getName();

  /**
   * Getter for rule description
   * @return rule description
   */

  String getDescription();

  /**
   * Determines if this rule evaluates
   * @return true if Rule evaluates and false otherwise
   * @param resource cloud resource to apply the rule on
   */

  boolean checkResource(Resource resource);

  /**
   * Checks if this rule applies
   * @param name name of the field to match
   * @return
   */

  boolean supports(String name);

  @Override
  default int compareTo(final Rule rule) {
    if (rule.getName().equalsIgnoreCase(getName()) && rule.getDescription().equalsIgnoreCase(getDescription())) {
      return 0;
    }

    return -1;
  }
}

File: swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/model/Rule.kt
/*
 * Copyright 2018 Netflix, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.netflix.spinnaker.swabbie.model

import com.netflix.spinnaker.config.ResourceTypeConfiguration.RuleDefinition

/**
 * A resource specific rule
 * If the rule finds the resource to be invalid, it will return a violation summary
 * Rules should be kept to simple logic and not perform any I/O operations
 */
interface Rule {
  fun <T : Resource> applicableForType(clazz: Class<T>): Boolean = false
  fun <T : Resource> apply(resource: T, ruleDefinition: RuleDefinition? = null): Result
  fun name(): String = this.javaClass.simpleName
}

data class Result(
  val summary: Summary?
)

data class Summary(
  val description: String,
  val ruleName: String
)
